### DISCUSSION TITLE
Writing Instrument Data to a File using the Data Storage App and the MSG type

### DISCUSSION THREAD

**Discussion by mc-bane** [nasa/cFS#282] [Q&A] [answered]
Somewhat of a followup to #278 

### Context
I am attempting to use cFS to record data from sensors on my system. Currently I have an app that interfaces with the sensors and forwards their telemetry to the cFS-GroundStation GUI. I would like to have a record of the readings from these sensors and the Data Storage application has some filtering functionality I am interested in exploiting for recording data. Since I have been working with a later version of cFS I've been updating various APIs on the DS app and can get it to compile, receive commands, and send some telemetry. 

Ultimately, I want a .csv file with a time column and sensor reading columns corresponding to a recorded time. 

### Question
Messages I write are not giving me data, and I'm not sure if it is a misunderstanding I may have made in the message IDs I am using, data corruption, or something else. 

Here is an output file for housekeeping after using hexedit to translate the binary to ascii. I thought I'd be able to read a timestamp or a count from the right column.
![image](https://user-images.githubusercontent.com/86273861/123285934-77564180-d4db-11eb-9ad8-0154af2155fc.png)

I'm quite stuck on this and would appreciate any insight.

**Comment by skliper** [comment, marked as answer]
You'll need to translate the binary data to csv.  hexedit just shows the ASCII representation of each byte, so data not written as ASCII won't show up as readable ASCII characters.  The file starts with the generic FS header:
```
typedef struct CFE_FS_Header
{
    uint32 ContentType;   /**< \brief Identifies the content type (='cFE1'=0x63464531)*/
    uint32 SubType;       /**< \brief Type of \c ContentType, if necessary */
                          /**< Standard SubType definitions can be found
                               \link #CFE_FS_SubType_ES_ERLOG here \endlink */
    uint32 Length;        /**< \brief Length of this header to support external processing */
    uint32 SpacecraftID;  /**< \brief Spacecraft that generated the file */
    uint32 ProcessorID;   /**< \brief Processor that generated the file */
    uint32 ApplicationID; /**< \brief Application that generated the file */


    uint32 TimeSeconds;    /**< \brief File creation timestamp (seconds) */
    uint32 TimeSubSeconds; /**< \brief File creation timestamp (sub-seconds) */


    char Description[CFE_FS_HDR_DESC_MAX_LEN]; /**< \brief File description */


} CFE_FS_Header_t;
```
Looking at your output you can see the Description element translated as "DS data storage file"... proceeded by 32 bytes of binary data.  The first element showing up as "cFE1" is because the uint32 binary value is set to translate to actual ASCII characters.

To translate, pick your favorite language to read in the binary values and print them out as ascii in your preferred format.
C-like proto code:
```
CFE_FS_Header_t CFE_FS_Header;
... (open the file)
read(file, &CFS_FS_Header, sizeof(CFS_FS_Header))); 
printf("0x%08x, "), CFS_FS_Header.ContentType);
... (print out the rest)
```

If you do this you'll see the first element printed as ascii: "0x63464531, ".  Or pick whatever print format you prefer (%u for decimal).
